File: evidence-service/tests/response-cache.test.ts
/**
 * Tests for ResponseCache — TTL-based in-memory cache.
 */

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { ResponseCache } from "../src/response-cache.js";

describe("ResponseCache", () => {
  let cache: ResponseCache;

  beforeEach(() => {
    cache = new ResponseCache();
  });

  it("returns null for missing keys", () => {
    expect(cache.get("expirations", "XLE")).toBeNull();
  });

  it("stores and retrieves within TTL", () => {
    cache.set("expirations", "XLE", [{ date: "2026-08-21", dte: 37 }], "2026-07-15T10:00:00Z");
    const result = cache.get<{ date: string; dte: number }[]>("expirations", "XLE");
    expect(result).not.toBeNull();
    expect(result!.data[0].date).toBe("2026-08-21");
    expect(result!.retrievedAt).toBe("2026-07-15T10:00:00Z");
  });

  it("returns null after TTL expires", () => {
    vi.useFakeTimers();
    cache.set("quote", "XLE", { price: 88.5, name: "Energy" }, "2026-07-15T10:00:00Z");

    // Advance past 60s quote TTL
    vi.advanceTimersByTime(61_000);

    expect(cache.get("quote", "XLE")).toBeNull();
    vi.useRealTimers();
  });

  it("different types have different TTLs", () => {
    vi.useFakeTimers();
    cache.set("expirations", "XLE", [], "now");
    cache.set("quote", "XLE", { price: 1 }, "now");

    // At 90 seconds: quote expired, expirations still valid
    vi.advanceTimersByTime(90_000);

    expect(cache.get("quote", "XLE")).toBeNull();
    expect(cache.get("expirations", "XLE")).not.toBeNull();
    vi.useRealTimers();
  });

  it("reports stats correctly", () => {
    cache.set("expirations", "A", [], "now");
    cache.set("expirations", "B", [], "now");
    cache.set("quote", "A", {}, "now");
    cache.set("chain", "A:2026-08-21", {}, "now");

    const stats = cache.stats();
    expect(stats.size).toBe(4);
    expect(stats.byType.expirations).toBe(2);
    expect(stats.byType.quote).toBe(1);
    expect(stats.byType.chain).toBe(1);
  });
});

File: evidence-service-java/src/test/java/com/wheelwright/evidence/provider/ResponseCacheTest.java
package com.wheelwright.evidence.provider;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * ResponseCache tests — TTL-based in-memory caching.
 * Behavioral parity with TypeScript response-cache.test.ts.
 */
class ResponseCacheTest {

    private ResponseCache cache;

    @BeforeEach
    void setUp() {
        cache = new ResponseCache();
    }

    @Test
    void returnsNullForMissingKeys() {
        assertNull(cache.get(ResponseCache.CacheType.EXPIRATIONS, "XLE"));
    }

    @Test
    void storesAndRetrievesWithinTtl() {
        cache.set(ResponseCache.CacheType.EXPIRATIONS, "XLE", "test-data", "2026-07-15T10:00:00Z");
        var result = cache.get(ResponseCache.CacheType.EXPIRATIONS, "XLE");
        assertNotNull(result);
        assertEquals("test-data", result.data());
        assertEquals("2026-07-15T10:00:00Z", result.retrievedAt());
    }

    @Test
    void differentKeysAreIndependent() {
        cache.set(ResponseCache.CacheType.EXPIRATIONS, "XLE", "xle-data", "now");
        cache.set(ResponseCache.CacheType.EXPIRATIONS, "SPY", "spy-data", "now");

        assertEquals("xle-data", cache.get(ResponseCache.CacheType.EXPIRATIONS, "XLE").data());
        assertEquals("spy-data", cache.get(ResponseCache.CacheType.EXPIRATIONS, "SPY").data());
    }

    @Test
    void differentTypesAreIndependent() {
        cache.set(ResponseCache.CacheType.EXPIRATIONS, "XLE", "exp-data", "now");
        cache.set(ResponseCache.CacheType.QUOTE, "XLE", "quote-data", "now");

        assertEquals("exp-data", cache.get(ResponseCache.CacheType.EXPIRATIONS, "XLE").data());
        assertEquals("quote-data", cache.get(ResponseCache.CacheType.QUOTE, "XLE").data());
    }

    @Test
    void reportsStatsCorrectly() {
        cache.set(ResponseCache.CacheType.EXPIRATIONS, "A", "a", "now");
        cache.set(ResponseCache.CacheType.EXPIRATIONS, "B", "b", "now");
        cache.set(ResponseCache.CacheType.QUOTE, "A", "qa", "now");
        cache.set(ResponseCache.CacheType.CHAIN, "A:2026-08-21", "ca", "now");

        ResponseCache.CacheStats stats = cache.stats();
        assertEquals(4, stats.size());
        assertEquals(2, stats.expirations());
        assertEquals(1, stats.quote());
        assertEquals(1, stats.chain());
    }
}
